piecewise/regressor.py

```diff
@@ -154,45 +165,84 @@ class SegmentTracker(object):
     """
 
     def __init__(self, segments):
-        self._ordered_segments = sorted(segments)
-        self._segment_set = set(segments)
+        # Assume segments are sorted
+        starts = np.fromiter((s.start_index for s in segments), np.intp, count=len(segments))
+
+        # One position for each original index.
+        # About 50% of this space is wasted, but this enables O(1) lookup and
+        # replacement by start_index
+        self._segments = np.empty(segments[-1].end_index, dtype=object)
+        self._segments[starts] = segments
+
+        # Valid mask.  As segments are merged, this mask is updated
+        self._valid = np.not_equal(self._segments, None)
+
+        # Previous neighbor lookup
+        self._prev = np.zeros_like(self._segments, dtype=np.intp)
+        self._prev[starts[1:]] = starts[:-1]
+
+        # Cached length.  Without this, we would need to count _valid every len()
+        self._len = len(segments)
 
     def __len__(self):
-        return len(self._ordered_segments)
+        return self._len
 
     def contains(self, segment):
         """ Returns True if segment is currently valid; False otherwise. """
-        return segment in self._segment_set
+        # segment at start_index has not been merged away and is still the same
+        return self._valid[segment.start_index] \
+            and self._segments[segment.start_index] is segment
```

Comment:
Using == raises "ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()" because cov_data is a numpy array.  Using reference equality works well with the algorithm.